src/hazardous/secret_stream/xchacha20poly1305.rs

```diff
@@ -0,0 +1,455 @@
+		new_key_and_inonce[CHACHA_KEYSIZE..]
+			.copy_from_slice(&self.nonce[SECRETSTREAM_XCHACHA20POLY1305_COUNTERBYTES..]);
+
+		chacha20_enc_in_place(
+			&self.get_key(),
+			&self.get_nonce(),
+			0,
+			&mut new_key_and_inonce,
+		)?;
+
+		self.key[..CHACHA_KEYSIZE].copy_from_slice(&new_key_and_inonce[..CHACHA_KEYSIZE]);
+		self.nonce[SECRETSTREAM_XCHACHA20POLY1305_COUNTERBYTES..]
+			.copy_from_slice(&new_key_and_inonce[CHACHA_KEYSIZE..]);
+
+		self.reset_counter();
+		Ok(())
+	}
+
+	#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+	/// TODO: Doc
+	/// //Ecryptes a message
+	pub fn push(
+		&mut self,
+		plaintext: &[u8],
+		ad: Option<&[u8]>,
+		dst_out: &mut [u8],
+		tag: Tag,
+	) -> Result<(), UnknownCryptoError> {
+		const BLOCKSIZE: usize = 64;
+		let mut block = [0u8; BLOCKSIZE];
+		let mut slen = [0u8; 8];
+		let pad = [0u8; 16];
+		let adlen = match ad {
+			Some(v) => v.len(),
+			None => 0,
+		};
+		let msglen = plaintext.len();
+		let cipherpos = core::mem::size_of::<Tag>();
+		let macpos = cipherpos + msglen;
+		if dst_out.len() < SECRETSTREAM_XCHACHA20POLY1305_ABYTES + msglen {
+			return Err(UnknownCryptoError);
+		}
+
+		chacha20_enc_in_place(&self.get_key(), &self.get_nonce(), 0, &mut block)?;
+		let mut poly = init(&OneTimeKey::from_slice(&block[..POLY1305_KEYSIZE])?);
+		use zeroize::Zeroize;
+		block.zeroize();
+		if adlen > 0 {
+			poly.update(ad.unwrap())?;
+		}
+		poly.update(&pad[..((16usize.overflowing_sub(adlen).0) & 15)])?;
+		block[0] = tag.bits();
+		chacha20_enc_in_place(&self.get_key(), &self.get_nonce(), 1, &mut block)?;
+		poly.update(&block)?;
+		dst_out[0] = block[0];
+
+		chacha20_enc(
+			&self.get_key(),
+			&self.get_nonce(),
+			2,
+			plaintext,
+			&mut dst_out[cipherpos..],
+		)?;
+		poly.update(&dst_out[cipherpos..(cipherpos + msglen)])?;
+		poly.update(
+			&pad[..((16usize
+				.overflowing_sub(BLOCKSIZE)
+				.0
+				.overflowing_add(msglen)
+				.0) & 15)],
+		)?;
+		store_u64_into_le(&[adlen as u64], &mut slen);
+		poly.update(&slen)?;
+		store_u64_into_le(&[BLOCKSIZE as u64 + msglen as u64], &mut slen);
+		poly.update(&slen)?;
+		let mac = poly.finalize()?;
+		dst_out[macpos..(macpos + mac.get_length())].copy_from_slice(mac.unprotected_as_bytes());
+		xor_buf8(self.get_inonce_mut(), &dst_out[macpos..macpos + 8]);
+		increment4(self.get_counter_mut());
+		if tag == Tag::REKEY
+			|| secure_cmp(
+				self.get_counter_mut(),
+				&[0u8; SECRETSTREAM_XCHACHA20POLY1305_COUNTERBYTES],
+			)
+			.is_ok()
+		{
+			self.rekey()?;
+		}
+		Ok(())
+	}
+
+	#[must_use = "SECURITY WARNING: Ignoring a Result can have real security implications."]
+	/// TODO: Doc
+	/// Decryptes a message
+	pub fn pull(
+		&mut self,
+		cipher: &[u8],
+		ad: Option<&[u8]>,
+		plaintext_out: &mut [u8],
+		tag_out: &mut Tag,
+	) -> Result<(), UnknownCryptoError> {
+		if cipher.len() < SECRETSTREAM_XCHACHA20POLY1305_ABYTES {
```

Comment:
If we check for `cipher.len() <= SECRETSTREAM_XCHACHA20POLY1305_ABYTES` instead, we also avoid processing empty ciphertexts.

Reply:
Do you happen to have a use-case that could help me understand when you'd want to send only tags that carry information and when encrypting streams? That is, scenarios where any such information can not be attached to a message?

Reply:
I think it should be generally possible to attach the tag to an message.

The question if empty ciphertexts should be allowed is more a generally design decision for this library. The rejection of empty texts can catch some user errors, but personally i would do this in the safe API and allow them in the hazardous API. This allows users that for some reasons want to do some operation on an empty text to do so.   Allowing empty texts should generally have no impact on memory or cryptographic safety.

Reply:
I would like to continue rejecting empty input for now at least. So we have to skip the plaintext encryption call if that is empty, but still authenticate the block that is then put in `dst_out`. Does that make any sense?